### PR TITLE
Add webhook

### PR DESCRIPTION
**Pull request by npmaile**
Updated to use previous stubs. 
Open as a draft to ensure compliance with code standards.
references issue #67

### CODE REVIEW

**Comment by npmaile** [comment]
The question I have left is if you want to fire the webhook and block until it's done or risk losing the alert in a go routine in the case that the program quits before it has a chance to send.

**Comment by vaibhav-kaushal** [comment]
@npmaile good question. Let me write that up the right response right away.

File: client/client.go [hunk 1]
```diff
@@ -307,6 +315,14 @@ func (c *Config) Alertf(message string, format ...any) {
 	l.MoreData = json.RawMessage("{}")
 	c.dispatchLogMessage(l)
 
+	if c.AlertWebhook != nil {
+		err := c.AlertWebhook(l)
+		if err != nil {
+			// log error sending webhook inline to avoid infinite loop and ignoring the message
+			c.Slogger.Log(context.Background(), barkslogger.LvlAlert, "unable to send webhook")
```
Comment (vaibhav-kaushal) [inline client/client.go:322]: We have to check if the Slogger is nil or not.

Comment (vaibhav-kaushal) [reply client/client.go:322]: and if it is, then we need to `fmt.Println` that entry.

**Comment by vaibhav-kaushal** [comment]
@npmaile check the above comment. I hope it makes sense!

File: client/client.go (hunk 1 above)
Comment (npmaile) [reply client/client.go]: Would you like a full models.Barklog printed (presumably as json) or just the string "unable to send alert"?

Comment (vaibhav-kaushal) [reply client/client.go]: The model struct can be used as a string. It formats itself. Look at this: https://github.com/techrail/bark/blob/3f5384a43eb7a41b587d5d7026865c981b20d4db/models/barklog.go#L63

**Comment by npmaile** [comment]
> @npmaile good question. Let me write that up the right response right away.
> 
> I think of it like this (there definitely can be better ways to look at it and I am open to knowing them) - from a very practical usage standpoint:
> 
>     1. A logging library must _never block on a network call, **ever**_. Most logging libraries are supposed to _print the logs_ to either a file or to the `stdout`. So that's out of the way - don't block, instead send it to a background goroutine.
> 
>     2. Now, we are talking about the `alert` level here - that's just one level lower than the highest level of `panic` level. So it is important that we get that alert across to the user so we should do what we can to send the alert to the user. So let's look at them.
> 
>     3. The first known way to lose the logs being emitted by the client is to kill the program too soon (before the network call could be made). So if your program is a very short lived program which just prints a lot of log messages and exits, you probably are _not in need of an alerting mechanism_. However, that kind of program can still be needed in production (for example, a job being executed by a background job execution system).
> 
>     4. It is thus imperative that we allow for both the behaviors -
>        a) Send them into background - useful for a long lived program like a server which needs to proceed with processing the next request.
>        b) Allow the client to block till the network call succeeds - in case of a program which needs to print the log and then just exist.
> 
>     5. And we need to put that control in the hands of the user.
> 
> 
> I think we can modify the `func (c *Config) Alert(message string)` to `func (c *Config) Alert(message string, block ... bool)`and then in the body of the function, we can do something like:
> 
> ```go
> if len(block) > 0 && block[0] == true {
>     err := c.Webhook(l)
>     // print the error as well as `l` in case of error and return
> } else {
>    go func() {
>      err := c.Webhook(l)
>       // print the error as well as `l` in case of error and return
>     }()
> }
> ```

I'm curious why using the variadic bool argument for blocking was chosen here instead of a standard bool. 
Also, would it make more sense to put the "alert failed to send" message into the "moreData" field of the log message itself, or should a new message be constructed to indicate the alert failure?

**Comment by vaibhav-kaushal** [comment]
Variadic because intuitively, I would just want to do a `log.Alert("something really important")` than `log.Alert("something really important", true)` every single time. But now that you have asked, I think that since Alerting is not a common level, especially with a webhook, we can actually make that argument a normal/standard bool. Hmm...!! Thanks for pointing that out.

I think it is good to do it something like:

```go
err := c.Webhook(l)
if err != nil {
  fmt.Printf("E#1LR1V1 - Webhook failed to send. Error: %v | Original Log Message: %v\n", err, l)
}
```

I hope that looks good?

File: client/client.go [hunk 2]
```diff
@@ -377,9 +425,9 @@ func (c *Config) Debugf(message string, format ...any) {
 	}
 }
 
-// func (c *Config) SetAlertWebhook(f webhook) {
-// 	c.AlertWebhook = f
-// }
+func (c *Config) SetAlertWebhook(f webhook, block bool) {
```
Comment (vaibhav-kaushal) [inline client/client.go:428]: We are not using `block` parameter here. It can be removed, I guess.

Comment (npmaile) [reply client/client.go:428]: oh shoot! I forgot to get rid of that

**Comment by vaibhav-kaushal** [comment]
@npmaile Have you tested the function? If not, I would have to first test it before I can actually merge the PR.

**Comment by npmaile** [comment]
> @npmaile Have you tested the function? If not, I would have to first test it before I can actually merge the PR.

yeah, I wrote up a little program (that seems to work) to kick the tires. I can do something a little more formal if you wish
```go
package main

import (
	"errors"
	"fmt"
	"time"

	"github.com/techrail/bark/client"
	"github.com/techrail/bark/models"
)

func main() {
	c := client.NewClient("http://localhost:18080", "A", "test", "test1", true, false)
	c.SetAlertWebhook(func(l models.BarkLog) error {
		fmt.Printf("alert #%s!", l.String())
		return nil
	})
	c.Alert("1", false)
	c.Alert("2", true)
	c.Alertf("3", false)
	c.Alertf("4", true)

	c.SetAlertWebhook(func(l models.BarkLog) error {
		return errors.New("simulated failure")
	})
	c.Alert("5", false)
	c.Alert("6", true)
	c.Alertf("7", true)
	c.Alertf("8", false)

	c2 := client.NewClient("http://localhost:18080", "A", "test", "test1", false, false)
	c2.SetAlertWebhook(func(l models.BarkLog) error {
		fmt.Println("alert!")
		fmt.Println(l.String())
		return nil
	})
	c2.Alert("9", false)
	c2.Alert("10", true)
	c2.Alertf("11", false)
	c2.Alertf("12", true)

	c2.SetAlertWebhook(func(l models.BarkLog) error {
		return errors.New(fmt.Sprintf("simulated failure #%s", l.String()))
	})
	c2.Alert("13", false)
	c2.Alert("14", true)
	c2.Alertf("15", true)
	c2.Alertf("16", false)

	c3 := client.NewClient("http://localhost:18080", "A", "test", "test1", false, false)
	c3.SetAlertWebhook(func(l models.BarkLog) error {
		time.Sleep(time.Second * 2)
		fmt.Printf("alert #%s\n", l.String())
		return nil
	})

	c3.Alertf("17", false)
	c3.Alertf("18", true)

	c4 := client.NewClient("http://localhost:18080", "A", "test", "test1", false, false)
	c4.SetAlertWebhook(func(l models.BarkLog) error {
		time.Sleep(time.Second * 2)
		return fmt.Errorf("simulated failure #%s", l.String())
	})

	c4.Alertf("19", false)
	c4.Alertf("20", true)
	time.Sleep(time.Second * 1)
}
```

**Comment by vaibhav-kaushal** [comment]
This is good. You can add this program to the `cmd/test-client-webhooks` directory (You'd have to create that one) and that should be fine for now.

**Pull request merged**